File: sources/nim/_11_parallel_sum.nim
# task 11 parallel_sum — expected output: 7500000075000000
# build: nim c -d:release -o:prog _11_parallel_sum.nim    run: ./prog
#
# Nim 2.0 enables thread support by default, so the build line needs no extra
# flag (add --threads:on only when building with an older compiler).
# Mechanism: std/typedthreads (Nim 2.0 has no std/threads module). Each worker
# gets its own fixed range plus a pointer to its slot of the shared result
# array; the main thread joins all four and sums the partials.

import std/typedthreads

type
  Work = tuple[lo, hi: int64, res: ptr int64]

proc worker(p: Work) {.thread.} =
  var acc: int64 = 0
  var i = p.lo
  while i < p.hi:
    case i mod 4
    of 0: inc acc
    of 1: acc += i
    of 2: acc += 2 * i
    of 3: acc += 3 * i
    else: discard
    inc i
  p.res[] = acc

var partials: array[4, int64]
var threads: array[4, Thread[Work]]
for t in 0 ..< 4:
  createThread(threads[t], worker,
    (lo: int64(t) * 25_000_000, hi: int64(t + 1) * 25_000_000,
     res: addr partials[t]))
joinThreads(threads)

var total: int64 = 0
for v in partials:
  total += v
echo total

File: sources/julia/11_parallel_sum.jl
# task 11 parallel_sum — expected output: 7500000075000000
# build: julia -t4 11_parallel_sum.jl    run: julia -t4 11_parallel_sum.jl
# deviation: Julia has no switch/match statement, so worker t uses the same if/elseif
# chain as task 02. Each thread owns one range and writes its own slot of `results`.

function partial(t::Int64)::Int64
    acc = Int64(0)
    lo = t * 25000000
    hi = lo + 25000000 - 1
    for i in lo:hi
        m = i % 4
        if m == 0
            acc += 1
        elseif m == 1
            acc += i
        elseif m == 2
            acc += 2 * i
        else
            acc += 3 * i
        end
    end
    return acc
end

function main()
    results = Vector{Int64}(undef, 4)
    Threads.@threads for t in 0:3
        results[t + 1] = partial(Int64(t))
    end
    println(sum(results))
end

main()
